RTCDataChannel close event does not fire if RTCPeerConnection closes

youennf opened the pull request:
#### e83716d173835cbf698bf91763d63a630776fa4e
<pre>
RTCDataChannel close event does not fire if RTCPeerConnection closes
<a href="https://rdar.apple.com/165617848">rdar://165617848</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=303052">https://bugs.webkit.org/show_bug.cgi?id=303052</a>

Reviewed by Jean-Yves Avenard.

Since closing the peer connection backend is severing the state updates of the data channel backend, we add the logic to close the data channel from RTCPeerConnection::close.
For that reason, we keep a list of channel identifiers in RTCPeerConnection.

We then call RTCDataChannelHandlerClient::peerConnectionIsClosing for each of this identifier.
This will do the following:
- If the data channel lives in main thread, it will call RTCDataChannel::peerConnectionIsClosing() directly.
- If the data channel lives in another process, it will call RTCDataChannel::peerConnectionIsClosing() directly.
- If the data channel lives in a worker, it will call RTCDataChannel::peerConnectionIsClosing() after hopping to the service worker thread.
- If the data channel is being transferred, it will call RTCDataChannel::peerConnectionIsClosing() which will make the newly transferred data channel moving to closed state.

Covered by added test.

Canonical link: <a href="https://commits.webkit.org/304859@main">https://commits.webkit.org/304859@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7a09742a84436b806d9981fe316eb5c4c821fbdc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136687 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/9046 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47974 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144415 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/89660 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/38349d02-727f-49ac-9a2c-faeba2e40c38) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/138559 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9747 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8892 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104525 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/76323 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/6b7bc2ad-2753-4735-b28d-43b142fed1c1) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139632 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/7123 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122476 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85364 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/9b0ad75d-2d44-446a-b18e-08fe2c0bcc44) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6767 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4455 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/5007 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/116085 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40664 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/147172 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8730 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/41237 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112880 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8748 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7347 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/113209 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28756 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6691 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118767 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/62863 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8778 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36821 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8499 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/72344 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8718 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8570 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/Modules/mediastream/RTCDataChannelRemoteSource.cpp:
```diff
@@ -36,22 +36,78 @@ namespace WebCore {
     , m_handler(WTFMove(handler))
     , m_connection(WTFMove(connection))
 {
+    ASSERT(isMainThread());
```
isMainRunLoop() ? as it's webkit

cdumez replied:
What? This feedback doesn't seem right. This file is in WebCore, does this code really run in the UIProcess (I doubt it).

jyavenard commented on Source/WebCore/Modules/mediastream/RTCDataChannelRemoteSource.cpp:
```diff
@@ -36,22 +36,78 @@ namespace WebCore {
+
+void RTCDataChannelRemoteSource::didReceiveStringData(const String& text)
+{
+    ASSERT(isMainThread());
```
ditto for all those in the file

cdumez replied:
What? This feedback doesn't seem right. This file is in WebCore, does this code really run in the UIProcess (I doubt it).

youennf committed:
RTCDataChannel close event does not fire if RTCPeerConnection closes
rdar://165617848
https://bugs.webkit.org/show_bug.cgi?id=303052

Reviewed by Jean-Yves Avenard.

Since closing the peer connection backend is severing the state updates of the data channel backend, we add the logic to close the data channel from RTCPeerConnection::close.
For that reason, we keep a list of channel identifiers in RTCPeerConnection.

We then call RTCDataChannelHandlerClient::peerConnectionIsClosing for each of this identifier.
This will do the following:
- If the data channel lives in main thread, it will call RTCDataChannel::peerConnectionIsClosing() directly.
- If the data channel lives in another process, it will call RTCDataChannel::peerConnectionIsClosing() directly.
- If the data channel lives in a worker, it will call RTCDataChannel::peerConnectionIsClosing() after hopping to the service worker thread.
- If the data channel is being transferred, it will call RTCDataChannel::peerConnectionIsClosing() which will make the newly transferred data channel moving to closed state.

Covered by added test.

Canonical link: https://commits.webkit.org/304859@main
